### ISSUE TITLE
unnecessary load (memory) error when accessing variable attributes with `return=`

### ISSUE THREAD

**Issue by karlmsmith** [NOAA-PMEL/Ferret#1585]
**Reported by andrew wittenberg on 11 Sep 2015 16:21 UTC**
Sometimes we want to access the attributes of a variable, and not the associated data -- e.g. when setting up new variables with titles derived from previous variables.  So the following is a problem:

```
 	NOAA/PMEL TMAP
 	FERRET v6.95 (rc2)
 	Linux 2.6.32-504.23.4.el6.x86_64 64-bit - 06/17/15
 	11-Sep-15 12:14     

yes? set mem/size=1
 Cached data cleared from memory
yes? let a = x[gx=1:1e6:1]
yes? let b = 1
yes? let/title="hi" c = a + b
yes? say `c,r=title`
 **ERROR: request exceeds memory setting: 1 Mwords were requested.
           *** NOTE: You can use SET MEMORY/SIZE=xxx to increase memory.
           *** NOTE: The "Memory use" section of the FERRET Users Guide has further tips.
```

Why is Ferret requesting memory (ostensibly to load the variable), when the title should be right there for the taking?

(In my application, I'm trying to define a bunch of variables, BEFORE the user sets a region to make the request(s) fit into memory.  This way they can change the region and re-evaluate, etc.)

Migrated-From: http://dunkel.pmel.noaa.gov/trac/ferret/ticket/2313

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 11 Sep 2015 17:06 UTC**
Indeed. 

The `RETURN=` machinery generally evaluates the expression as that's the only way to determine some of the return items, such as details about the grid.  However a bunch of them can be determined without that, including title, units, the missing-flag and other attributes.

`RETURN=defined` already skips the evaluation step. I'll work through which other return items could also bypass full evaluation.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 14 Sep 2015 17:44 UTC**
When the RETURN requests something that is one of the attributes that are part of the definition of the variable; or an attribute of a file variable from a dataset, it can be returned immediately without evaluating the expression.

units, title, bad, user_scale, user_offset fall into this category. For file variables, dtype, xmod, tmod, isdepth, t0, f0, calendar, dset, dsetnum, dsetpath, dsettitle, nc_scale, nc_offset can also be returned without loading anything.


fixed in repl_exprns.F

**Comment by karlmsmith** [comment]
**Comment by andrew.wittenberg on 23 Jun 2017 19:47 UTC**
It seems there's still a loose end here, when we simply multiply by a scalar:

```
 	NOAA/PMEL TMAP
 	FERRET v7.12 (optimized)
 	Linux 2.6.32-696.1.1.el6.x86_64 64-bit - 05/16/17
 	23-Jun-17 15:42     

yes? set mem/size=1
yes? def ax/t=1-jan-0001:1-jan-1000:1/unit=year tax
yes? let a = t[+ x[gx=1:1000:1](gt=tax]) + y[gy=1:1000:1]
yes? let one = 1
yes? say `a,r=cal`
 !-> MESSAGE/CONTINUE GREGORIAN
GREGORIAN
yes? say `a*one,r=cal`
 **ERROR: request exceeds memory setting
    To fulfill this request would exceed the current SET MEMORY/SIZE= limit of 1 megawords
    At the moment that the memory limit was reached
    memory was committed as follows:
        - to objects used in computation: : 2000 (0.2%)
    The size of the requested object was: : 1000000 (100%)
```

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 23 Jun 2017 21:11 UTC**
There seems to be something peculiar about handling the scalar variable "let one = 1".

Try this example, multiplying first by a fixed number, then by an expression on a dataset grid,

```
yes? set mem/siz=1
yes? def ax/t=1-jan-0001:1-jan-1000:1/unit=year tax
yes? let a = t[+ x[gx=1:1000:1](gt=tax]) + y[say `6*a,r=cal`
 !-> MESSAGE/CONTINUE GREGORIAN
GREGORIAN

yes? use coads_climatology
yes? let s1= sst[x=@ave,y=1,L=@max](gy=1:1000:1]

yes?)
yes? say `s1*a,r=cal`
           *** NOTE: Ambiguous coordinates on T axis: S1*A
 !-> MESSAGE/CONTINUE GREGORIAN
GREGORIAN
```

I suspect it's something about the grid of a scalar being the "ABSTRACT" grid:

```
yes? let one = 1
yes? sh grid one
    GRID ABSTRACT
 name       axis              # pts   start                end                 subset
 ABSTRACT  X             99999999 r   1                    1.E+08              full
 ABSTRACT  Y             99999999 r   1                    1.E+08              full
 ABSTRACT  Z             99999999 r   1                    1.E+08              full
 ABSTRACT  T             99999999 r   1                    1.E+08              full
 ABSTRACT  E             99999999 r   1                    1.E+08              full
 ABSTRACT  F             99999999 r   1                    1.E+08              full
```

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 23 Jun 2017 21:19 UTC**
More clues,

```
yes? set mem/siz=1
yes? def ax/t=1-jan-0001:1-jan-1000:1/unit=year tax
yes? let a = t[+ x[gx=1:1000:1](gt=tax]) + y[def axis/x=1:1:1 xax1
yes? let onex = 1+0*x[gx=xax1](gy=1:1000:1]
yes?)
yes? say `onex*a,r=cal`
 !-> MESSAGE/CONTINUE GREGORIAN
GREGORIAN
```
or, similarly,
```
yes? set mem/siz=1
yes? def ax/t=1-jan-0001:1-jan-1000:1/unit=year tax
yes? let a = t[+ x[gx=1:1000:1](gt=tax]) + y[let onex = 1+0*x[x=1:1:1](gy=1:1000:1]
yes?)
yes? say `onex*a,r=cal`
 !-> MESSAGE/CONTINUE GREGORIAN
GREGORIAN
```

I'm not suggesting these as workarounds, but as a pointer to what's causing this. The script in comment 3 should work.

**Comment by karlmsmith** [comment]
**Comment by ansley.b.manke on 30 Jun 2017 22:03 UTC**
This is fixed.  When reconciling all the grids within an expression, if part of the expression is a variable that is a scalar, do not inherit the abstract axes of the scalar variable's grid into grid directions that are otherwise normal.  Done in merge_grid.

This is an improvement on the fix to ticket #1602, which also came out of behavior of the grid of a constant such as LET one = 1
